### PR TITLE
[AIR] Add `TorchVisionPreprocessor`

### LINKED ISSUE #30403: [Datasets] [AIR] Add `TorchVisionPreprocessor`
### Description

Add a preprocessor like

```
TorchVisionTransformer(transform: Callable[[], torch.Tensor], columns: list[str])
```

that applies [TorchVision transforms](https://pytorch.org/vision/stable/transforms.html).

### Use case

To apply a simple `TorchVision` transform like:

```
transform = Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]) 
```

You need to something like this:

```
def map_fn(batch: dict[str, np.ndarray]) -> dict[str, np.ndarray]:
    return {"image": np.array([transform(image).numpy() for image in batch], dtype=object)}
```

which is complicated.

With a `TorchVisionTransformer`, you could instead do:

```
TorchVisionTransformer(transform, columns="image")
```

which is simpler.

### PR DESCRIPTION
**Pull request by bveeramani**
Signed-off-by: Balaji <balaji@anyscale.com>

**Depends on**:
- [x] #30448 
- [x] #30514 

## Why are these changes needed?

**tl;dr**: Users can't figure out how to apply TorchVision transforms. This PR introduces an abstraction that makes it easy.

To apply a TorchVision transform, you need to do something like this:
```
def preprocess(batch: dict[str, np.ndarray]) -> dict[str, np.ndarray]:
    return {"image": np.array(transform(image).numpy() for image in batch["image"])}

BatchMapper(preprocess, batch_format="numpy")
```

There're many unobvious things:
* You need to set `batch_format="numpy"` (`batch_format` is a required argument, but it's not obvious if you should use `"numpy"` or `"pandas"`).
* You need to individually apply the transformation to each image in the batch. 
* You need to convert transformed images back to `ndarray`s.
* You need to wrap the batch of transformed images in an `ndarray`.
* You need to return a dictionary. 

This PR introduces a preprocessor `TorchVisionPreprocessor` that abstracts away the complexity:

```
TorchVisionPreprocessor(columns=["image"], transform=transform)
```

## Related issue number

Closes #30403 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/pandas_block.py [hunk 1]
```diff
@@ -197,18 +197,27 @@ def to_numpy(
...
                     f"Cannot find column {column}, available columns: "
                     f"{self._table.columns.tolist()}"
                 )
+
         arrays = []
         for column in columns:
             arrays.append(self._table[column].to_numpy())
-        if len(arrays) == 1:
+
+        if should_be_single_ndarray:
```
Comment (Yard1) [inline python/ray/data/_internal/pandas_block.py:220]: should we assert here too?

**Review by Yard1 — APPROVED** [review]
The API & preprocessor look good to me, will defer on Data changes to Data team!

File: python/ray/data/_internal/pandas_block.py (hunk 1 above)
Comment (bveeramani) [reply python/ray/data/_internal/pandas_block.py:220]: Data changes are from the stacked PRs:
* https://github.com/ray-project/ray/pull/30448
* https://github.com/ray-project/ray/pull/30514

This PR should only touch AIR-related stuff.

**Comment by clarkzinzow** [comment]
@bveeramani What are the next steps for this PR? I think this was blocked by a tensor extension bug, but the fix should be merged into master.

**Comment by bveeramani** [comment]
> @bveeramani What are the next steps for this PR? I think this was blocked by a tensor extension bug, but the fix should be merged into master.

@clarkzinzow I think I just need to resolve merge conflicts and maybe add a couple more tests.

**Comment by amogkam** [comment]
Some torchvision transforms work on a batched input. Should we allow users to also specify a separate batched_transform for better performance?

**Comment by bveeramani** [comment]
> Some torchvision transforms work on a batched input. Should we allow users to also specify a separate batched_transform for better performance?
> 
> See our batch prediction benchmarks for an example

How do you imagine this would work? 

In pure Torch,  you'd use typically use `torchvision.transforms.Compose`. So, I imagined people would do something like:
```python
transform = transforms.Compose(
    ToTensor(),
    CenterCrop(224),
    Normalize(mean=..., std=...)
)

preprocessor = TorchVisionPreprocessor(transform)
```

To support batched transforms, would you need to do something like:
```python
preprocessor = Chain(
    TorchVisionPreprocessor(ToTensor()),
    TorchVisionPreprocessor(CenterCrop(224), batched=True),
    TorchVisionPreprocessor(Normalize(mean=..., std=...), batched=True)
)
```

**Comment by bveeramani** [comment]
> @bveeramani Why couldn't you do a single `transforms.Compose`d batch transform given to a single `TorchVisionPreprocessor`?

Not all transforms support batches. Like, you could compose some batch-supporting transforms with transforms that don't support batches.

How would we implement this? If we straightforwardly pass a batch of data to `Compose`, the program would error if the on something like `ToTensor`.

**Comment by clarkzinzow** [comment]
I was thinking that we could make it opt-in at the `TorchVisionPreprocessor` constructor, i.e.
```python
transform = transforms.Compose(
    CenterCrop(224, batched=True),
    Normalize(mean=..., std=..., batched=True)
)
preprocessor = TorchVisionPreprocessor(transform, batched=True)
```

**Comment by amogkam** [comment]
API could be something like this:

```
single_transform = transforms.ToTensor()
batch_transform = transforms.Compose([CenterCrop(...), Normalize(...)])
preprocessor = TorchVisionPreprocessor(transform=single_transform, batch_transform=batch_transform)
```

Single preprocessor that can accept both options. `batch_transform` arg is Optional.

File: python/ray/data/preprocessors/torch.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            if self._batched:
+                return self._fn(batch).numpy()
+            return np.array([self._fn(array).numpy() for array in batch])
+
+        if isinstance(np_data, dict):
+            outputs = {
+                column: transform(batch)
+                for column, batch in np_data.items()
+                if column in self._columns
+            }
```
Comment (amogkam) [inline python/ray/data/preprocessors/torch.py:90]: this currently drops all the columns that are not used by this preprocessor? we should keep those around so that they can continue to be used by downstream preprocessors right?

**Review by amogkam — COMMENTED** [review]
Awesome, looks great so far! Left a couple comments.

Can we also update all the existing code that currently manually use torchvision preprocessors?

File: python/ray/data/preprocessors/torch.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            f"transform={self._fn!r})"
+        )
+
+    def _transform_numpy(
+        self, np_data: Union["np.ndarray", Dict[str, "np.ndarray"]]
+    ) -> Union["np.ndarray", Dict[str, "np.ndarray"]]:
+        def transform(batch: np.ndarray) -> np.ndarray:
+            if self._batched:
+                return self._fn(batch).numpy()
+            return np.array([self._fn(array).numpy() for array in batch])
```
[the remark this answers is not in the document]
Comment (bveeramani) [reply python/ray/data/preprocessors/torch.py:83]: I don't think so. It's complex to check and it might be overly restrictive. 

I've seen nested transforms like:

```
Compose(
    Compose(
        ToTensor(),
        SpamTransform(),
    ),
    HamTransform()
```

To implement the check, we'd need to recursively type-check if something is a `Compose`. 

Also, you could do something like

```
Lambda(lambda array: torch.as_tensor(array).permute((2, 0, 1))
```

and we wouldn't want to raise an error.

File: python/ray/data/preprocessors/torch.py (hunk 2 above)
Comment (bveeramani) [reply python/ray/data/preprocessors/torch.py]: Yeah, good catch. Fixed.

**Comment by bveeramani** [comment]
Failing tests are because of `FileNotFoundError: Couldn't find file at https://www.dropbox.com/s/1pzkadrvffbqw6o/train.txt?dl=1` in `test_huggingface`

**Review by amogkam — APPROVED** [review]
Thanks!

**Review by richardliaw — APPROVED** [review]
